### PR TITLE
Allow RWC runner to use default library in built/local/lib.d.ts

### PR DESCRIPTION
**Pull request by yuit**

### CODE REVIEW

**Comment by danquirk** [comment]
lgtm. Do we have any tests that use this yet to verify it's working?

**Comment by yuit** [comment]
@danquirk I do. But it is RWC tests. I am going to send out code review for that separately

File: src/harness/harness.ts [hunk 1]
```diff
@@ -1703,6 +1703,14 @@ module Harness {
...
 
+    export function getDefaultLibraryFile(): { unitName: string, content: string } {
+        var libFile = Harness.userSpecifiedroot + Harness.libFolder + "/" + "lib.d.ts";
+        return {
+            unitName: libFile,
+            content: IO.readFile(libFile)
+        }
+    }
+
     if (Error) (<any>Error).stackTraceLimit = 1;
```
Comment (DanielRosenwasser) [inline src/harness/harness.ts]: Curlies

File: src/harness/rwcRunner.ts [hunk 2]
```diff
@@ -43,6 +44,10 @@ module RWC {
                 baselineOpts = undefined;
                 baseName = undefined;
                 currentDirectory = undefined;
+                // useDefaultLibFile is a flag specified in the json object to indicate whether to use built/local/lib.d.ts
+                // or to use lib.d.ts inside the json object. If the flag is true or undefine, use the build/local/lib.d.ts
+                // otherwise use the lib.d.ts from the json object.
```
Comment (DanielRosenwasser) [inline src/harness/rwcRunner.ts]: undefine -> undefined

Also, that is very weird; we're sure that it's *`true`* or `undefined`? Not `false` or `undefined`?

File: src/harness/rwcRunner.ts [hunk 3]
```diff
@@ -51,32 +56,52 @@ module RWC {
...
-                        if (libFile) {
-                            inputFiles.push(getHarnessCompilerInputUnit(libFile));
-                        }
-                    }
-
+                    // Add files to compilation
                     ts.forEach(ioLog.filesRead, fileRead => {
+                        // Check if the file is already added into the set of input files.
                         var resolvedPath = ts.normalizeSlashes(ts.sys.resolvePath(fileRead.path));
                         var inInputList = ts.forEach(inputFiles, inputFile=> inputFile.unitName === resolvedPath);
```
Comment (DanielRosenwasser) [inline src/harness/rwcRunner.ts]: Space before `=>`

File: src/harness/rwcRunner.ts [hunk 4]
```diff
@@ -51,32 +56,52 @@ module RWC {
...
-                        if (!inInputList) {
```
Comment (DanielRosenwasser) [inline src/harness/rwcRunner.ts]: Just do

```TypeScript
if (inInputList) {
    return;
}
```

or if you convert the forEach to a for..of, use

```TypeScript
if (inInputList) {
    continue;
}
```

File: src/harness/rwcRunner.ts [hunk 5]
```diff
@@ -51,32 +56,52 @@ module RWC {
...
+                                // If useDefaultLibFile is true, we will use built/local/lib.d.ts
+                                // instead of the provided lib.d.ts from json object.
+                                // Majority of RWC code will be using built/local/lib.d.ts instead of
+                                // lib.d.ts inside json file. However, some RWC cases will still use
+                                // their own version of lib.d.ts because they have customized lib.d.ts
+                                if (useDefaultLibFile) {
+                                    inputFiles.push(Harness.getDefaultLibraryFile());
+                                }
+                                else {
+                                    inputFiles.push(getHarnessCompilerInputUnit(fileRead.path));
```
Comment (DanielRosenwasser) [inline src/harness/rwcRunner.ts]: Might be cleaner as a ternary, experiment.

File: src/harness/rwcRunner.ts [hunk 6]
```diff
@@ -115,9 +140,7 @@ module RWC {
 
             it('has the expected declaration file content', () => {
                 Harness.Baseline.runBaseline('has the expected declaration file content', baseName + '.d.ts', () => {
-                    if (compilerResult.errors.length || !compilerResult.declFilesCode.length) {
-                        return null;
-                    }
+                    if (!compilerResult.declFilesCode.length) {                        return null;                    }
```
Comment (DanielRosenwasser) [inline src/harness/rwcRunner.ts]: Personally would prefer length > 0

**Pull request merged**